unit_tests/test_redis_aggregator.py

```diff
@@ -0,0 +1,238 @@
+# pylint: disable=missing-function-docstring, protected-access
+import json
+import uuid
+from unittest.mock import patch, PropertyMock, MagicMock
+import pytest
+
+from gsy_e_sdk.redis_aggregator import RedisAggregator, RedisAggregatorAPIException
+
+TEST_AGGREGATOR_NAME = "TestAgg"
+
+TEST_AREA_NAME = "TestArea"
+TEST_AREA_MAPPING = {TEST_AREA_NAME: [str(uuid.uuid4())]}
+
+TEST_TRANSACTION_ID = str(uuid.uuid4())
+TEST_AGGREGATOR_UUID = str(uuid.uuid4())
+
+TEST_DEVICE_UUID_1 = str(uuid.uuid4())
+TEST_DEVICE_UUID_2 = str(uuid.uuid4())
+TEST_BATCH_COMMAND = "TEST_BATCH_COMMAND"
+
+TEST_BATCH_COMMAND_DICT = {
+    TEST_DEVICE_UUID_1: [TEST_BATCH_COMMAND],
+    TEST_DEVICE_UUID_2: [TEST_BATCH_COMMAND]
+}
+
+TEST_TRANSACTION_ID_RESPONSE = "TEST_TRANSACTION_ID_RESPONSE"
+TEST_TRANSACTION_ID_BUFFER = {
+    TEST_TRANSACTION_ID: {"transaction_id": TEST_TRANSACTION_ID,
+                          "response": TEST_TRANSACTION_ID_RESPONSE}
+}
+
+TEST_RESPONSE = "TEST_RESPONSE"
+TEST_TRANSACTION_ID_RESPONSE_BUFFER = {TEST_TRANSACTION_ID: TEST_RESPONSE}
+
+
+@pytest.fixture(autouse=True)
+def fixture_mock_connections(mocker):
+    mocker.patch("gsy_e_sdk.redis_aggregator.StrictRedis")
+    mocker.patch("gsy_e_sdk.redis_aggregator.ThreadPoolExecutor")
+
+
+@pytest.fixture(name="mock_grid_fee_calculation")
+def fixture_mock_grid_fee_calculation(mocker):
+    """Mock GridFeeCalculation class."""
+    mocker.patch("gsy_e_sdk.redis_aggregator.GridFeeCalculation")
+
+
+@pytest.fixture(name="mock_transaction_id_and_timeout_blocking")
+def fixture_mock_transaction_id_and_timeout_blocking(mocker):
+    mocker.patch("gsy_e_sdk.redis_aggregator.uuid.uuid4",
+                 return_value=TEST_TRANSACTION_ID)
+    mocker.patch("gsy_e_sdk.redis_aggregator.wait_until_timeout_blocking")
+
+
+@pytest.fixture(name="mock_client_command_buffer_attributes")
+def fixture_mock_client_command_buffer(mocker):
+    mocker.patch("gsy_e_sdk.redis_aggregator.ClientCommandBuffer.buffer_length",
+                 new_callable=PropertyMock, return_value=3)
+    mocker.patch("gsy_e_sdk.redis_aggregator.ClientCommandBuffer.execute_batch",
+                 return_value=TEST_BATCH_COMMAND_DICT)
+
+
+@pytest.fixture(name="aggregator")
+def fixture_aggregator(mocker):
+    mocker.patch("gsy_e_sdk.redis_aggregator.RedisAggregator._create_aggregator")
+    return RedisAggregator(aggregator_name=TEST_AGGREGATOR_NAME)
+
+
+class TestRedisAggregatorConstructor:
+    """Includes tests for the constructor of the RedisAggregator class."""
```

Comment:
Are you suggesting here that it would be ok if I grouped several assertions in a single test for the constructor of the class? Or how this should be in order to not override the "one assertion per test" rule.
Or could you please clarify how you would do it?

Reply:
Well I would personally aggregate these assertions in as single test case, I'm not a fan of the one assertion per test methodology as long as all the assertions are related/similar/simple and relevant in the context of the test case performed

Reply:
To be honest, I think that I actually am against this one assertion per test case rule, I find it unreasonable and it would exponentially prolong our test modules, imo if a test is related to one specific event (in our case initialization of the aggregator) it's totally okay to write several assertions for this particular event

Reply:
The one assertion per test makes sense IMO, but it's not necessary all the time. In simpler cases, I think this is not needed.

Longer version: the rationale behind the one-assertion "rule" is that each unit test should fail for only one reason, because this will save us from spending time debugging the reason why the test failed. From a different perspective: each test should only be set up for one single scenario (the one we're testing). If we test two different situations, this is not good because we can't immediately know which part of our setup (mocks etc) caused the test to fail. Using one assertion makes things easier because it forces us to keep the setup atomic/simple for our unit test.
Also, the one-assertion rule works best for methods that are kept short and that respect the single-responsibility principle.

Reply:
When we have several scenarios (perhaps conditional statements in the code), we could split the one single test method into several scenario based tests where we append the event/action/scenario description in the name of test.
This doesn't necessarily mean that we should only have one assertion for these tests, because each node (conditional statement) in our codebase can lead to multiple changes elsewhere (so we should be able to track all of these changes)
In general, I agree that the one assertion per test CAN make sense for methods that follow the single responsibility principle, especially ones that change or return states (we can assert the values returned/set)